### DISCUSSION TITLE
New expressions:  PrevX(), PrevY(), MovementDirection(), DistanceMoved()

### DISCUSSION THREAD

**Discussion by tristanbob** [4ian/GDevelop#3642] [Enhancement ideas (and their technical discussions)]
## Description

Provide expressions that describe how the object moved in the last frame:

object.PrevX()  OR  object.X(-1)  
- Previous X position 

object.PrevY()  OR  object.Y(-1)   
- Previous Y position

object.MovementDirection()
- Direction moved from position in the previous frame to current position

object.DistanceMoved()
- Distance moved from position in the previous frame to current position

## Alternatives considered

"object.ForceDistance()" and "object.ForceAngle()" sound exactly what I wanted, but they only take into account movement caused by forces.  This means it does not include movement initiated by platformer/top-down/physics behaviors, mouse dragging, manual position changes, etc.

"object.Direction()" sounds like the right thing, but it appears to be functionally similar to "object.Angle()".

These expressions can be manually calculated by saving the last frame position and comparing it to the current frame (like FaceForward does) but I think these expressions should be included by default.

**Comment by 4ian** [comment]
Do you think these should be included by default, or can we back these into a "MovementTracker" behavior? Having a behavior has the interesting aspect that we would not consume memory (to store the previous position) for all objects, and we could also expand this "tracker" to include more features (inspect the movement like the distance moved over the last X seconds?)

**Comment by arthuro555** [comment]
[StateSaveExt.zip](https://github.com/4ian/GDevelop/files/6749167/StateSaveExt.zip)
I made a minimal extension implementing the gist of this feature request

**Comment by 4ian** [comment]
Nice! You think we could make this an extension on the registry?

**Comment by arthuro555** [comment]
Maybe though that draft was made quickly more as a PoC than a real solution, and still needs to be worked on. I'm a bit busy with another extension right now but I can see later if I can look further into it if no one else is interested in finishing it.
